api4/resolver.go

```diff
@@ -0,0 +1,281 @@
+	}
+	return c.App.ClientConfigWithComputed(), nil
+}
+
+// match with api4.getClientLicense
+func (r *resolver) License(ctx context.Context) (model.StringMap, error) {
+	c, err := getCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionReadLicenseInformation) {
+		return c.App.Srv().ClientLicense(), nil
+	}
+	return c.App.Srv().GetSanitizedClientLicense(), nil
+}
+
+// match with api4.getTeamMembersForUser for teamID=""
+// and api4.getTeamMember for teamID != ""
+func (r *resolver) TeamMembers(ctx context.Context, args struct {
+	UserID string
+	TeamID string
+}) ([]*teamMember, error) {
+	c, err := getCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if args.UserID == model.Me {
+		args.UserID = c.AppContext.Session().UserId
+	}
+
+	if !c.App.SessionHasPermissionToUser(*c.AppContext.Session(), args.UserID) && !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionReadOtherUsersTeams) {
+		c.SetPermissionError(model.PermissionReadOtherUsersTeams)
+		return nil, c.Err
+	}
+
+	canSee, appErr := c.App.UserCanSeeOtherUser(c.AppContext.Session().UserId, args.UserID)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	if !canSee {
+		c.SetPermissionError(model.PermissionViewMembers)
+		return nil, c.Err
+	}
+
+	if args.TeamID != "" {
+		if !c.App.SessionHasPermissionToTeam(*c.AppContext.Session(), args.TeamID, model.PermissionViewTeam) {
+			c.SetPermissionError(model.PermissionViewTeam)
+			return nil, c.Err
+		}
+
+		tm, appErr2 := c.App.GetTeamMember(args.TeamID, args.UserID)
+		if appErr2 != nil {
+			return nil, appErr2
+		}
+
+		return []*teamMember{{*tm}}, nil
+	}
+
+	members, appErr := c.App.GetTeamMembersForUser(args.UserID)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	// Convert to the wrapper format.
+	res := make([]*teamMember, 0, len(members))
+	for _, tm := range members {
+		res = append(res, &teamMember{*tm})
+	}
+
+	return res, nil
+}
+
+func (*resolver) ChannelsLeft(ctx context.Context, args struct {
+	UserID string
+	Since  float64
+}) ([]string, error) {
+	c, err := getCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if args.UserID == model.Me {
+		args.UserID = c.AppContext.Session().UserId
+	}
+
+	if !c.App.SessionHasPermissionToUser(*c.AppContext.Session(), args.UserID) {
+		c.SetPermissionError(model.PermissionEditOtherUsers)
+		return nil, c.Err
+	}
+
+	return c.App.Srv().Store.ChannelMemberHistory().GetChannelsLeftSince(args.UserID, int64(args.Since))
+}
+
+// match with api4.getChannelMember
+func (*resolver) ChannelMembers(ctx context.Context, args struct {
+	UserID       string
+	ChannelID    string
+	First        int32
+	After        string
+	LastUpdateAt float64
+}) ([]*channelMember, error) {
+	c, err := getCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if args.UserID == model.Me {
+		args.UserID = c.AppContext.Session().UserId
+	}
+
+	// If it's a single channel
+	if args.ChannelID != "" {
```

Comment:
This is missing permission checks for the case where `args.ChannelID == ""`. Pass in an arbitrary user ID and get info about their channel memberships.

Reply:
Ohh dear me, what a bug. :cry:  Amazing eye there @jupenur :pray:

Reply:
I'm actually pretty happy with how GraphQL works even when there's a bug like this one. Because the permission checks still kick in when you hit any other resolver, the impact is very limited. With an equivalent bug in the REST APIs, the impact would be similarly limited thanks to the scoping of the individual endpoints, but there's one major difference: with REST you could still read the channel IDs of the channels you don't have access to. That would allow you to uniquely identify the different channel memberships -- which would then allow you to accurately track an arbitrary user joining and leaving private channels. With this bug you can still do some of that tracking, but to a much lesser extent because you can't reliably distinguish one channel membership from another.

So for a positive angle on this, it was actually valuable for me to see a bug like this play out. It gives me confidence that adopting GraphQL could actually end up improving our security resilience 🙂